File: stack/0503_next_greater_element_ii/0503_next_greater_element_ii.go
package main 

import (
	"fmt"
)

func nextGreaterElements(nums []int) []int {
	result := make([]int, len(nums))
	stack := make([]int, len(nums))
	top := -1

	for i := range result {
		result[i] = -1
	}

	for i := 0; i<len(nums) * 2 ;i++ {
		num := nums[i % len(nums)]
		for top > -1 && nums[stack[top]] < num {
			result[stack[top]] = num
			top--
		}
		if i < len(nums) {
			top++
			stack[top] = i
		}
	}
	return result
}

func main() {
	nums := []int{1,2,1}
	res := nextGreaterElements(nums)
	fmt.Println(res)
}

File: stack/0503_next_greater_element_ii/0503_next_greater_element_ii.py

class Solution(object):
    def nextGreaterElements(self, nums):
        stack = []
        result = [-1]*len(nums)
        for enum, n in enumerate(nums):
            while stack and stack[-1][0] < n:
                result[stack[-1][1]] = n
                stack.pop()
            stack.append((n, enum))
        print result, stack
        if stack:
            for enum, n in enumerate(nums):
                while stack and stack[-1][0] < n:
                    result[stack[-1][1]] = n
                    stack.pop()

        return result

nums = [1,2,1]
res = Solution().nextGreaterElements(nums)
print(res)